### PR TITLE
aether: init at 2.0.0-dev.15

### PR DESCRIPTION
**Pull request by maxhille**
The changes in this PR add myself to the maintainer list and the Aether P2P client as a package.

about Aether: https://getaether.net/

###### Motivation for this change

No package yes. People were asking for it (https://github.com/NixOS/nixpkgs/issues/66414)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by maxhille** [comment]
Whoever reviews this:

This is my first new package contribution, so I don't expect to have done everything right. I don't mind direct critique. <3

File: pkgs/applications/networking/aether/default.nix
Comment (roberth) [inline pkgs/applications/networking/aether/default.nix:14]: These two aren't used in the derivation. See `libPath` below. You can use a let binding instead.

**Review by roberth — COMMENTED** [review]
You're off to a good start.

Please update the commit messages to conform to the guidelines mentioned in the pr template. You could use `git rebase -i` and the `r` (reword) rebase command to achieve this.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 130105` at 232ed92c run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>aether</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/applications/networking/aether/default.nix:50:3:
```
   |
50 |   installPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`agpl3` is a deprecated license, please check if project uses `agpl3Plus` or `agpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/networking/aether/default.nix:85:5:
```
   |
85 |     license = licenses.agpl3;
   |     ^
```
</li>
  </ul>
</details>

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:73]:
```suggestion
    exec = "Aether";;
```

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:14]:
```suggestion
let
  binaryName = "AetherP2P";
in
stdenv.mkDerivation rec {
  pname = "aether";
  version = "2.0.0-dev15";
```

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:17]:
```suggestion
    url = "https://static.getaether.net/Releases/Aether-${version}/2011262249.19338c93/linux/Aether-${version}%2B2011262249.19338c93.tar.gz";
```

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:26]: that goes definitely into native.

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:85]: Do you know if the license is agpl3Only or agpl3Plus?

**Comment by maxhille** [comment]
@SuperSandro2000 @roberth How exactly is the merge process going to happen (ie. who pushes the merge button)?

I am asking because I want to address your review comments with single commits (otherwise Github gets confused IIRC) but still squash those into the "package init" commit before in lands on master.

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:85]: I do not think the Aether authors thought about that. They only copied the LICENSE file in verbatim into their repo root. I checked a couple of the source files and they did not include a version notice at all. I read through the relevant sections of the AGPL text and as far as I understand it, the raw license text does not imply only or plus.

Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:85]: https://github.com/aethereans/aether-app

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:35]: When I remove that, I get ```/home/mh/.nix-profile/bin/aetherp2p: error while loading shared libraries: libxkbcommon.so.0: cannot open shared object file: No such file or directory```. So I guess it is necessary. Of course I have no clue what `nativeBuildInputs` are because the Nix build system just confuses me most of the time.

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:73]: Why do you suggest that?

BTW, your suggestion would not work as the executable is called `AetherP2P` but I guess that was just by mistake.

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:26]: What do you mean by that? Is there something I should do?

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix]: I addressed this, please resolve if you are fine with it.

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix]: I addressed this, please resolve if you are fine with it.

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix]: I addressed this, please resolve if you are fine with it.

Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix]: This should be addresses as part of https://github.com/NixOS/nixpkgs/pull/130105/commits/647d9c9d77e505f4d526352ba3b7dcd2dde64e0c

Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix]: Please resolve when happy

File: pkgs/applications/networking/aether/default.nix
Comment (roberth) [reply pkgs/applications/networking/aether/default.nix:23]: I think he just means `autoPatchelfHook` is in the right place. I agree.

File: pkgs/applications/networking/aether/default.nix
Comment (roberth) [reply pkgs/applications/networking/aether/default.nix:32]: `buildInputs` and `nativeBuildInputs` are unnecessarily confusing. I hope to improve this with my [suggestion here](https://github.com/NixOS/nixpkgs/issues/28327#issuecomment-879815573).

My best explanation is:
 - `buildInputs` is for stuff that is "like" your own build and can become part of the output
 - `nativeBuildInputs` is for stuff that runs during the derivation's build

However, the distinction isn't currently enforced, and I didn't know that `nativeBuildInputs` libraries end up treated like `buildInputs` libraries when `strictDeps = false` (the default).

All of this is mostly relevant for cross compilation, which isn't actually relevant here, but code like this could cause confusion in other readers, which is why I'm asking.

And I didn't ask clearly. I guess what needs to happen is, could you move the libraries to a `buildInputs` attribute?
Programs that run during the build can stay in `nativeBuildInputs`.

File: pkgs/applications/networking/aether/default.nix
Comment (roberth) [reply pkgs/applications/networking/aether/default.nix:82]: From the license

> If the Program does not specify a version number of the GNU Affero General Public License, you may choose any version ever published by the Free Software Foundation.

However, the package.json refers to the file, which itself _is_ a version. This is hilarious.

We should ask them. Do we have a boilerplate for that? It would be nice to have a good explanation of the question and its relevance to this community and others.

Comment (roberth) [reply pkgs/applications/networking/aether/default.nix:82]: If anyone knows of earlier cases, they can share their experience here https://github.com/jtojnar/nixpkgs-hammering/issues/118

**Comment by 06kellyjac** [comment]
A member can either merge it as it is in the PR or squash merge it for you.
It's preferable for you to squash them yourself and forcepush to the PR.

Just remember that adding yourself as a maintainer and init of the package should be 2 separate commits :)

File: maintainers/maintainer-list.nix
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:6607]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:9]:
```suggestion
}:

stdenv.mkDerivation rec {
```

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/networking/aether/default.nix:23]: Yeah and almost everything else is buildInputs.

**Comment by maxhille** [comment]
@roberth @SuperSandro2000 as far as I can tell I addressed the suggestions. I also added some more code because I had to get the app icon from their git repository. I hope the force push did not throw off Github too much (happens some time).

Regarding the license I wrote a request at their bugtracker[1] (My post is still pending moderation).

[1] https://meta.getaether.net/top/all

File: maintainers/maintainer-list.nix
Comment (maxhille) [reply maintainers/maintainer-list.nix:6607]: I reworded the commit. please resolve when happy

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:55]: Thanks for the explanation. Please re-check

Comment (roberth) [reply pkgs/applications/networking/aether/default.nix:55]: lgtm

File: pkgs/applications/networking/aether/default.nix
Comment (roberth) [reply pkgs/applications/networking/aether/default.nix:110]: Request has been made. Current value is sufficient for now.

**Review by roberth — APPROVED** [review]
Looks good to me.

Regarding the use of string constants, I don't expect most of them stand the test of an actual change happening, and the indirection could be slightly harmful. `version` is a notable exception. I don't feel very strongly about it though, so I wouldn't delay a merge for it.

Maybe @SuperSandro2000 wants to have another look. Otherwise, good for merge afaic.

**Comment by maxhille** [comment]
@roberth @SuperSandro2000 I squashed the package commits together, so from my POV this is ready to merge.

Regarding the strings constant I admit it is a bit pointless as of now, eg. you will find `aether-2.0.0-dev.15` aswell as `aether-2.0.0-dev15` (dot missing). I think the project does not have some sort of formal release process and as such abstracting would probably be a waste of time right now.

If the Aether project stays alive I assume we will rework this derivation to build from source at some point anyway.

**Comment by roberth** [comment]
@maxhille I agree.

The build did fail to fetch the archive though. That's hard to test locally because fixed-output derivation outputs are always cached. You can delete the cached paths from your store using 

```
nix-store --delete $(nix-store -q --outputs $(nix-instantiate -A aether.srcs))
```

If gc fails, you may have to remove gc roots:

```
nix-store -q --roots $(nix-instantiate -A aether.srcs)
```

**Comment by maxhille** [comment]
> The build did fail to fetch the archive though. That's hard to test locally because fixed-output derivation outputs are always cached. You can delete the cached paths from your store using

Turns out that missing dot was my fault somehow. Fixed now
 
> ```
> nix-store --delete $(nix-store -q --outputs $(nix-instantiate -A aether.srcs))
> ```
> 
> If deletion fails, you may have to remove gc roots:
> 
> ```
> nix-store -q --roots $(nix-instantiate -A aether.srcs)
> ```

Thanks for the pointer.

FYI I had to do `nix-store -q --roots $(nix-instantiate -E 'with import <nixpkgs> {}; callPackage ./default.nix {}' -A srcs)` otherwise `nix-instantiate` would fail with `error: cannot auto-call a function that has an argument without a default value ('autoPatchelfHook')`.

**Comment by roberth** [comment]
Ah that makes sense when run in the aether directory. I assumed to run from the nixpkgs root.

**Comment by maxhille** [comment]
> Ah that makes sense when run in the aether directory. I assumed to run from the nixpkgs root.

I guess that also explains why I had to remove the `aether` qualifier in the attribute.  Never stop learning - thanks for the help, I appreciate it!

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:109]:
```suggestion
    license = licenses.agpl3Only;
```

You don't need to ask upstream about this. If there is no mention of ``or later`` outside of the standard license text it is only.

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:76]: Use copyDesktopItems instead.

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:102]: Use copyDesktopItems instead.

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:99]: Why do we need opt? Can we use the standard directories instead?

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:40]:
```suggestion
  buildPhase = ''
    convert Linux-Windows-App-Icon.png -resize 512x512 aether.png
  '';
```

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:35]:
```suggestion
  postUnpack = ''
    cp aether-app-53b6c8b/aether-core/aether/client/src/app/ext_dep/images/Linux-Windows-App-Icon.png ${sourceRoot}
  '';
```

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:17]:
```suggestion
```

File: pkgs/applications/networking/aether/default.nix
Comment (benhylau) [reply pkgs/applications/networking/aether/default.nix:109]: Hello, I'm from Aether. I have [clarified the license to be **AGPL-3.0-only**](https://github.com/aethereans/aether-app/commit/d054809743b03e0167299abb680a622888763a71) after speaking with @nehbit. Thanks!

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:18]: I have to rename that because the URL contains a `%` which results in an error otherwise.

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:72]: Can you give a reason why I should do that?

Unfortunately I have serious trouble to find the sources for the imported functions, so I don't know more about `copyDesktopItems` other than that it does not work as a substitute because some argument is missing...

Also please note that the original tarball does not come with a desktop file, so we _have_ to create one. As said I don't know what `copyDesktopItem` does but I assume is copies existing files?

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:98]: See https://github.com/NixOS/nixpkgs/pull/130105#discussion_r673241053

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix:95]: Maybe we need it, maybe not. As stated earlier I took what the `discord` package is doing. I actually expect other stuff to be non-optimal here, too. But also keep in mind that this package wants to be built from source at some point (in the future) anyway.

**Comment by maxhille** [comment]
I added some formatting suggestions and changed the license to `agpl3Only`.

**Comment by maxhille** [comment]
@roberth Seem like a check for some inferior, proprietary OS got stuck. Is there something I can do about it?

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:18]:
```suggestion
      # % in the filename causes an error
      name = "aether-tarball.tar.gz";
```

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:22]: Please use fetchFromGitHub.

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:36]:
```suggestion
    convert Linux-Windows-App-Icon.png -resize 512x512 aether.png   '';
```

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:32]:
```suggestion
    cp aether-app-53b6c8b/aether-core/aether/client/src/app/ext_dep/images/Linux-Windows-App-Icon.png ${sourceRoot}
  '';
```

**Comment by roberth** [comment]
> @roberth Seem like a check for some inferior, proprietary OS got stuck. Is there something I can do about it?

The problem is with our tooling. To be honest, I don't know the exact status of the ofborg x86_64-darwin builders, but I'm pretty sure it's not because of the operating system.

You can ignore the darwin commit status because you didn't package for it and `meta.platforms` is correct.

> some inferior, proprietary OS

Please don't make a fool of yourself.

Also, complacency is a real problem for open source if "it" wants to compete in the OS arena. It's not the root cause though and I won't blame any individual for it.

But hey, let's enjoy ourselves :)

**Comment by matthew-piziak** [comment]
@maxhille Thanks for your persistence here in diligently updating to feedback. It's looking like a great first package.

File: pkgs/applications/networking/aether/default.nix
Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix]: Thanks for the explanation.

In contrast to the example, I had to add `runHook postInstall` to the `installPhase` block, otherwise `copyDesktopItems` would just not execute.

Again I have no idea why that is...

**Comment by maxhille** [comment]
@SuperSandro2000 Can you revisit the suggestions / changes I made?

File: pkgs/applications/networking/aether/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/aether/default.nix:40]:
```suggestion

  buildPhase = ''
    convert source/aether-core/aether/client/src/app/ext_dep/images/Linux-Windows-App-Icon.png -resize 512x512 aether.png
```

Comment (maxhille) [reply pkgs/applications/networking/aether/default.nix]: done. had to add relative ``../source`` though.

**Comment by nehbit** [comment]
Hi all — maintainer of Aether P2P here. I've been out travelling for the past few weeks, so I'm just seeing all this. Great to see this work out. 

Let me know if you have any questions and/or requests. Once this is done and working, I'd be happy to add NixOS as officially supported and point it your way.

**Comment by SuperSandro2000** [comment]
In a couple of days it should be available on the unstable channel.

**Comment by matthew-piziak** [comment]
I wanted to test this, so using this config I created myself a `pkgs.aether-branch`:
```nix
let fetchWithConfig = src: import (fetchTarball src) { config = config.nixpkgs.config; };
in {
nixpkgs = {
  config = {
    allowUnfree = true;
    packageOverrides = _: {
      unstable = fetchWithConfig "https://github.com/NixOS/nixpkgs/archive/nixos-unstable-small.tar.gz";
      aether-branch = fetchWithConfig "https://github.com/NixOS/nixpkgs/archive/pull/130105/head.tar.gz";
    };
  };
};}
```
…and then I tried to install `pkgs.aether-branch.aether`, but I got `A definition for option environment.systemPackages.[definition 20-entry 1] is not of type package. Definition values:`, and then the message just ends. Is `aether` the correct package name?

**Comment by Danie10** [comment]
@nehbit is there a link to the dev16 release?

**Pull request merged**
